### PR TITLE
[WEB-5577] Send token to Lambda for authorization

### PR DESCRIPTION
**Pull request by mzhodzik**
[WEB-5577](https://bowery.atlassian.net/browse/WEB-5577)

### CODE REVIEW

File: src/scripts/background.js [hunk 1]
```diff
@@ -314,66 +314,66 @@ const postUserActivity = async (requestBody) => {
...
                 'Content-Type': 'application/json',
                 'Authorization': `Bearer ${idToken}`
             },
             body: JSON.stringify(requestBody)
         });
-    } catch(error) {
-        if (error.code === 403) {
+
+        if (!response.ok && response.status === 403) {
             authHelper.runAuthProcess();
```
Comment (MakaryHryshkin) [inline src/scripts/background.js]: await?

Comment (VeranikaVasilchykava) [reply src/scripts/background.js]: yep, you are right

File: src/manifest.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
...
...0CY7qhnXrPTolT4+JKM1hnUYqWtjKLlGtjhcnVJXzwI3+07NhYZ1+XGqPVWH0fRuTmM45yjFgndx/4si4/n4dIwImcz+T1K7BTnZThq3HelbS5u6Zg195W52YqYThaSQmQZfGFt2+A7+H6SVPX7UdU+ek4hM77Ftq5yTPcVOLKzCDIyDDGUkG1LKXNZ7vZRHJVWDfeu73yD2mu2n8CMdqpzXdF1OA4+7EOS2ZrwDn6vzYDb2uvUw+rnaCPRv6FYCapDQ8luiDbA2YtkDQvAYiTq0xNYm4x9nH52Lt58avGPLM5fWeXujg55mDDk2JbOQKBgQC8VnU9jF7IAt9m3DdWSs2oHCxj/+Plomoba9bEZ13auCWzIf3flnzx0PfUCZ+p28/W0+GxAJFOdSY88vQmhKddpt9yZqIl3ZAnoxm1NTn/R01/5PYkL5GKSIIl9z+htQ1SaCFDaJZGXGXd8eoSY8T606xqnacUui7aEkaObtYInQKBgQC3V4joVPl4JlAI/lICBecMmFvCkcqv6yWMdaY36NcbI/j2Z0fNDhiar+wUM+ycB7oltqpf6mt8gJSoxr9PW/Q/GRDROiymcyNZPWs/3A8PzwDmp/aqIEhyn2L52egnNbjvQxzrWE3aFCDp6KejOKY9POeZuoEPhBu7c9VDKED7uwKBgGM4MR+xuaUpTkREM6r0UZAE96XfA1LQ05GKqGRkCF/Xr5TFjZejEkPa0RUV1NWJhd8pVLcJE4P/15s4a0N3kyyVYxZD15yXMwCOvlZmqh1AlSWRm9LIdKYjON2DejRYy6HkVdHLYY/D0a6xaZ0yjm2N4kFwuLsEsvYTtrQG1B9JAoGAMKp3mNQVdCMXU4vOUOMoNVur37W7S37rdKE2Adk4X+foid9YW8HGHilcTPCLec9DRgd6qgMbjOzEiRhYZ8RMzQLmR+NdPsGJPILlGQpcpSGZZ6BmBTqUgHh2WUPSL7Y+5KO35XCHGcAGDJDJDrZWyiKEh3XOFERI6DUTAsSIm0kCgYEAus+vEG90iKirYCA8smdoxiUBnSu0PbUUouoxP9Hjtlelw189z07r6yzyIhzVIlKtXgX++Gl+TTazbK3ZFWcpC0JnoFpvGrnHrESohU6+hHZIxG3HPGWK4nyb0E2pTkhbFNEa2OQmUS+tVMB0awcgYyads/j6xYQLir0YcI+Axfw=",
```
Comment (mrromro) [inline src/manifest.json:19]: Hoooly cookies...

File: src/scripts/authHelper.js [hunk 3]
```diff
@@ -1,68 +1,86 @@
...
-    const email =
-      (await this._getEmailFromProfile()) ||
-      (await this._getEmailFromStorage()) ||
-      (await this._getEmailFromGoogleOAuth());
-
-    if (email) {
-      await this._saveEmail(email);
-      chrome.browserAction.setPopup({ popup: "../index.html" });
+    this._authAttempts++;
+    if (this._authAttempts > 3) return;
```
Comment (mrromro) [inline src/scripts/authHelper.js:11]: What should we do if we run out of attempts? It seems that we are doomed in this case.

**Review by mrromro — APPROVED** [review]
Nicely done.

File: src/scripts/background.js [hunk 4]
```diff
@@ -310,64 +310,80 @@ const mapTime = (item) => {
...
-        });
+        const response = await postUserActivityHandler(requestBody);
+
+        if (!response.ok && response.status === 403) {
+            throw new AuthenticationError(AUTH_ERROR_MSG);
+        }
     } catch(error) {
         console.error(error);
+        if (error instanceof AuthenticationError) {
+            await authenticationErrorHandler(requestBody);
```
Comment (MakaryHryshkin) [inline src/scripts/background.js]: handleAuthError?

File: src/scripts/authHelper.js (hunk 3 above)
Comment (MakaryHryshkin) [reply src/scripts/authHelper.js:11]: @mzhodzik let's discuss it once you are back

File: src/scripts/authHelper.js [hunk 5]
```diff
@@ -74,25 +92,58 @@ class AuthHelper {
...
+              resolve(code || null);
             }
           }
         }
       );
     });
   }
+
+  _getIdAndRefreshTokens(code) {
+    return new Promise(async (resolve, reject) => {
```
Comment (MakaryHryshkin) [inline src/scripts/authHelper.js:108]: not sure why we need promise here, async/away should be fine.

File: src/scripts/authHelper.js [hunk 6]
```diff
@@ -74,25 +92,58 @@ class AuthHelper {
...
+
+        resolve(responseJson);
+      } else {
+        resolve(null);
+      }
+    });
+  }
+
+  _getRefreshedIdToken(refreshToken) {
+    return new Promise(async (resolve, reject) => {
```
Comment (MakaryHryshkin) [inline src/scripts/authHelper.js:130]: same for promise here

File: src/scripts/authenticationError.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+"use strict";
```
Comment (MakaryHryshkin) [inline src/scripts/authenticationError.js]: can we put it in errors folder?

File: src/scripts/authHelper.js [hunk 8]
```diff
@@ -74,25 +92,58 @@ class AuthHelper {
...
+            redirect_uri: `https://${chrome.runtime.id}.chromiumapp.org/`,
+            grant_type: "refresh_token",
+          }),
+          method: "POST",
+        });
+        const responseJson = await response.json();
+
+        resolve(responseJson);
+      } else {
+        resolve(null);
```
Comment (MakaryHryshkin) [inline src/scripts/authHelper.js:146]: Do we need to resolve it with null?

File: src/scripts/authHelper.js [hunk 9]
```diff
@@ -1,68 +1,86 @@
...
+      if (idTokenExpirationDate > Date.now()) {
+        this._authAttempts = 0;
+        chrome.browserAction.setPopup({ popup: "../index.html" });
+      } else {
+        const { id_token, expires_in } = await this._getRefreshedIdToken(
+          refreshToken
+        );
+        this._saveIdTokenData(id_token, expires_in);
+
+        this.runAuthProcess();
```
Comment (MakaryHryshkin) [inline src/scripts/authHelper.js]: should this have await 😢 ?

File: src/scripts/common.js [hunk 10]
```diff
@@ -55,6 +58,7 @@ var SETTINGS_VIEW_TIME_IN_BADGE = 'view_time_in_badge';
 var SETTINGS_BLOCK_DEFERRAL = 'view_block_deferral';
 var SETTINGS_SHOW_HINT = 'show_hint';
 
+const CLIENT_SECRET = "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"
```
Comment (MakaryHryshkin) [inline src/scripts/common.js:62]: are we ok with this to be hardcoded?

**Review by MakaryHryshkin — APPROVED** [review]
Nice work, left a few comments

**Pull request merged**